File: apps/web/src/services/duckduckgo-search.ts
export interface SearchResult {
  title: string;
  snippet: string;
  url: string;
  domain: string;
}

/**
 * 100% Free, zero-API-key web search client using DuckDuckGo.
 */
export async function searchDuckDuckGo(
  query: string,
  maxResults = 5,
): Promise<SearchResult[]> {
  const cleanQuery = query.trim();
  if (!cleanQuery) return [];

  const url = `https://html.duckduckgo.com/html/?q=${encodeURIComponent(cleanQuery)}`;
  const results: SearchResult[] = [];

  try {
    const res = await fetch(url, {
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
        Accept:
          'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
      },
      signal: AbortSignal.timeout(10000),
    });

    if (!res.ok) {
      console.warn(`DuckDuckGo search failed with status ${res.status}`);
      return [];
    }

    const html = await res.text();
    const regex =
      /<a class="result__url"[^>]*href="([^"]+)"[\s\S]*?<a class="result__snippet"[^>]*>([\s\S]*?)<\/a>/g;
    let match: RegExpExecArray | null;

    while ((match = regex.exec(html)) !== null && results.length < maxResults) {
      const rawUrl = match[1];
      let cleanUrl = rawUrl;
      const uddgMatch = rawUrl.match(/uddg=([^&]+)/);
      if (uddgMatch) {
        cleanUrl = decodeURIComponent(uddgMatch[1]);
      }

      const snippet = match[2]
        .replace(/<[^>]+>/g, '')
        .replace(/&quot;/g, '"')
        .replace(/&#x27;/g, "'")
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .trim();

      const domainMatch = cleanUrl.match(/^https?:\/\/([^/]+)/);
      const domain = domainMatch ? domainMatch[1] : 'web';

      if (snippet && cleanUrl.startsWith('http')) {
        results.push({
          title: domain,
          snippet,
          url: cleanUrl,
          domain,
        });
      }
    }
  } catch (error) {
    console.warn('DuckDuckGo search error:', error);
  }

  return results;
}

File: backend/data_sources/duckduckgo_search.py
import logging
import re
import urllib.parse
from typing import Any
import httpx

logger = logging.getLogger(__name__)


class DuckDuckGoSearch:
    """Zero-configuration, 100% free web search client powered by DuckDuckGo."""

    def __init__(self, timeout: float = 12.0):
        self.timeout = timeout
        self.headers = {
            'User-Agent': (
                'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                'AppleWebKit/537.36 (KHTML, like Gecko) '
                'Chrome/124.0.0.0 Safari/537.36'
            ),
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
        }

    async def search(self, query: str, max_results: int = 5) -> list[dict[str, Any]]:
        clean_query = query.strip()
        if not clean_query:
            return []

        url = f'https://html.duckduckgo.com/html/?q={urllib.parse.quote(clean_query)}'
        results = []

        try:
            async with httpx.AsyncClient(headers=self.headers, timeout=self.timeout, follow_redirects=True) as client:
                resp = await client.get(url)
                if resp.status_code != 200:
                    logger.warning("duckduckgo_search_failed status=%d", resp.status_code)
                    return []

                html = resp.text
                pattern = re.compile(
                    r'<a class="result__url"[^>]*href="([^"]+)"[\s\S]*?<a class="result__snippet"[^>]*>([\s\S]*?)</a>'
                )

                for match in pattern.finditer(html):
                    raw_url, raw_snip = match.group(1), match.group(2)
                    uddg = re.search(r'uddg=([^&]+)', raw_url)
                    dest_url = urllib.parse.unquote(uddg.group(1)) if uddg else raw_url

                    # Clean html tags & entities from snippet
                    clean_snip = (
                        re.sub(r'<[^>]+>', '', raw_snip)
                        .replace('&quot;', '"')
                        .replace('&#x27;', "'")
                        .replace('&amp;', '&')
                        .replace('&lt;', '<')
                        .replace('&gt;', '>')
                        .strip()
                    )

                    # Extract domain
                    domain_match = re.search(r'https?://([^/]+)', dest_url)
                    domain = domain_match.group(1) if domain_match else 'web'

                    if clean_snip and dest_url.startswith('http'):
                        results.append({
                            'url': dest_url,
                            'snippet': clean_snip,
                            'domain': domain,
                        })

                    if len(results) >= max_results:
                        break

        except Exception as e:
            logger.warning("duckduckgo_search_exception query=%s error=%s", clean_query, str(e))

        return results
